data/impl/RecordSet.js

```diff
@@ -54,63 +51,124 @@ export class RecordSet {
+            if (parent && !filter(parent)) return false;
+            return true;
+        };
+
+        return this.applyFilter(filter);
     }
 
     /**
-     * Return a version of this recordset with a record replaced.
-     * NOTE: Currently replacing a record at the root is the only supported operation.
+     * Return a version of this recordset with additional records added or updated.
      *
-     * @param {Record} oldRecord
-     * @param {Record} newRecord
+     * @param {Object[]} rawData -  raw data for records to be added or updated.
      * @return {RecordSet}
      */
-    updateRecord(oldRecord, newRecord) {
-        const newRoots = clone(this.roots),
-            index = findIndex(newRoots, {id: oldRecord.id});
+    updateData(rawData) {
+        const newRecords = this.createRecords(rawData),
+            existingRecords = new Map(this.records);
+
+        newRecords.forEach((newRecord, id) => {
+            const currRecord = existingRecords.get(id);
+            if (!currRecord || !currRecord.isEqual(newRecord)) {
+                existingRecords.set(id, newRecord);
+            }
+        });
 
-        if (index < 0) throw XH.exception(`Cannot find Record to update: id = ${oldRecord.id}`);
+        return new RecordSet(this.store, existingRecords);
+    }
 
-        newRecord.children = oldRecord.children;
-        newRoots[index] = newRecord;
 
-        return new RecordSet(newRoots);
+    //-----------------------------------
+    // Implementatation
+    //-----------------------------------
+    createRecords(rawData) {
+        const ret = new Map();
+        rawData.forEach(raw => this.createRecord(raw, ret, null));
+        return ret;
     }
 
+    createRecord(raw, records, parent) {
+        const {store} = this,
+            {idSpec} = store,
+            idGen = isString(idSpec) ? r => r[idSpec] : idSpec;
+
+        if (store.processRawData) store.processRawData(raw);
+
+        raw.id = idGen(raw);
+        throwIf(
+            isNil(raw.id),
+            "Record has a null/undefined ID. Use the 'LocalStore.idSpec' config to resolve a unique ID for each record."
+        );
+        throwIf(
+            records.has(raw.id),
+            `ID ${raw.id} is not unique. Use the 'LocalStore.idSpec' config to resolve a unique ID for each record.`
+        );
+        const rec = new Record({raw, parent, store});
+        records.set(rec.id, rec);
+        if (raw.children) {
+            raw.children.forEach(rawChild => this.createRecord(rawChild, records, rec));
+        }
+    }
     
     //------------------
     // Implementation
     // ------------------
-    gatherAllRecords(records, list = [], map = new Map()) {
-        records.forEach(r => {
-            list.push(r);
-            map.set(r.id, r);
-            if (!isEmpty(r.children)) {
-                this.gatherAllRecords(r.children, list, map);
-            }
-        });
-        return {list, map};
+    getParent(rec) {
```

Comment:
Any reason not to just have a `parent` prop on Record now that it know about its store?

Reply:
yes, can add a (dynamic) getter for that.  Good idea